agent.go

```diff
@@ -315,7 +290,160 @@ func (a *Agent) watchHealthChecks(updateCh chan api.HealthChecks) {
 			continue
 		}
 
+		ourChecks := make(api.HealthChecks, 0)
+		for _, c := range checks {
+			if ourNodes[c.Node] {
+				ourChecks = append(ourChecks, c)
+			}
+		}
+
 		opts.WaitIndex = meta.LastIndex
-		updateCh <- checks
+		updateCh <- ourChecks
+	}
+}
+
+// computeWatchedNodes watches both the list of registered ESM instances and the list of
+// external nodes registered in Consul and decides which nodes the local agent should be
+// in charge of in a deterministic way.
```

Comment:
This is only determinisitic assuming the input at every node is consistent.

Consul's catalog is strongly consistent, but the view you get from watching at different nodes will not be reliably. We expect they will soon converge though so I think this is OK.

But for example this is not going to maintain any guarantee that there is only ever 1 ESM that thinks it's is responsible for a given service instance, nor that at all times at least one ESM does.

I _think_ it bounds the amount of time either case can be true under the assumption that any "unhealthy" ESM node that does not have the latest service nodes or ESM node list will eventually fail it's health checks and so be taken out causing other nodes to re-shuffle.

I wouldn't be confident though without lots and lots of thought and modelling that this can't have bugs under realistic circumstances where there is no ESM that thinks it's responsible for some services which seems a bit dangerous.

Having not thought too hard yet, it _seems_ like it would be much easier to reason about a scheme that actually encodes the mapping of service -> monitoring ESM in the state store and have the leader update that as it sees ESMs/service instances come and go. It should be equivalent on the happy path but having non-leader ESM's just watch one key and update their config when it changes seems easier to reason about than each watching two separate things and handling any ordering of updates between them.

I could be convinced though it's mostly an initial reaction that this scheme seems pretty hard to reason about the possible failure modes and how bad they might be (i.e. nothing monitoring some instances).

Reply:
I was originally going to keep the leader lock and have it write to a KV entry for each agent, but I think this is just simpler and has the same resilience/failure modes. In either case, the ESM needs a working connection to Consul in order to both get its nodes to watch and update their health, so there won't be a chance for two ESM agents to be watching the same nodes for any significant length of time.

The one case where ESMs could overlap for a short period of time is when adding or removing a new external node or ESM, but this update should propagate so fast that I don't think it'll be an issue in practice. Having a leader that writes each ESM's watch list to the state store or KV entries has the same issue, as the ESMs aren't guaranteed to all get the update at the same time. I think watching two separate things (external nodes and ESM instances) is fine here as we don't need to care about the ordering of any updates - the watched nodes are just a function of the last update on each.

The only way I see around this would be to have the leader process use something like ACLs for each ESM and update the permissions on every rebalance to prevent the wrong agent from writing any health check state, but that both requires ACLs to be enabled and seems overkill to avoid something that doesn't seem like it will be an issue in the first place.

Reply:
Yeah that's more or less what I thought. I guess it's OK I just find it hard to have confidence in all the emergent behaviours this might have or the assumptions that might be violated under "grey failures".

I think this is "simple" in the sense of less code to write, but I don't know that it will be very simple to reason about how things get into buggy states later.

Most similar systems I've seen prefer to elect a coordinator (leader) node and then have it explicitly assign work and monitor that it's being done. Kafka does this to assign stream partitions to brokers for example. The key difference there is that the leader directly coordinates the workers and is responsible for making sure that all the work is assigned to something healthy at all times.

I don't want to over-engineer this, maybe it will work well enough as it is. 🤷‍♂️